File: use insert sort

espr14 opened the pull request:
Insert sort is same or faster than bubble sort while remaining sorting stability.

espr14 committed:
File: use insert sort

leptun committed:
Fix build

leptun committed:
Implement timer elapsed function


also fix build

leptun committed:
Remove quicksort, shellsort and add performance benchmark

leptun commented:
Force pushed onto the latest MK3 branch and some other improvements.

Some stats I did on an old SD card I had around (time is in ms):
```
sortingSpeedtestTimer:2295 - insert time /
sortingSpeedtestTimer:3056 - bobl time /

sortingSpeedtestTimer:1407 - insert alpha /
sortingSpeedtestTimer:6594 - bobl alpha /

sortingSpeedtestTimer:38 - insert time /old
sortingSpeedtestTimer:76 - bobl time /old

sortingSpeedtestTimer:31 - insert alpha /old
sortingSpeedtestTimer:37 - bobl alpha /old
```

From this short test, I can say the perfomance is greatly improved. More testing should be done with more files in the folder (if I can find my test image I was using a few years ago)

leptun commented:
In some situations this algorithm is considerably slower than the old one:
```
sortingSpeedtestTimer:8993 - insert time syncthing/waldi-prusa/SD_menu/SD.img.xz/prvni/druha/treti
sortingSpeedtestTimer:558 - bobl time syncthing/waldi-prusa/SD_menu/SD.img.xz/prvni/druha/treti

sortingSpeedtestTimer:1246 - insert alpha syncthing/waldi-prusa/SD_menu/SD.img.xz/prvni/druha/treti
sortingSpeedtestTimer:558 - bobl alpha syncthing/waldi-prusa/SD_menu/SD.img.xz/prvni/druha/treti
```

This happened because the files were already sorted both for the time based sorting and for the alphabetical sorting (and there were more than 100 files).

leptun committed:
Remove postponed sorting of elements

Now that the position is a uint16_t instead of uint32_t, it is simpler to just sort the positions in place without using the uint8_t indices. Also, this approach is considerably lighter on the stack usage and it also removes a delay after the sorting happens

leptun committed:
Use the sd cache for positions even when not sorting

leptun committed:
Disable sorting speedtest
